### PR TITLE
mon: fix "OSDs marked OUT wrongly after monitor failover"

### PR DESCRIPTION
**Pull request by sysnote**
mon: update mon(peon)'s down_pending_out when osd up

Fixes: http://tracker.ceph.com/issues/17719
Signed-off-by: Dong Wu archer.wudong@gmail.com

### CODE REVIEW

**Comment by tchaikov** [comment]
could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes. the title of your commit now reads "Fixes: http://tracker.ceph.com/issues/17719"

**Comment by tchaikov** [comment]
@jecluis do you want to take a look also?

**Comment by sysnote** [comment]
@tchaikov updated.

**Comment by tchaikov** [comment]
@sysnote could you squash these three commits into a single one and repush (`git push -f`)?

**Comment by sysnote** [comment]
@tchaikov updated, thanks for your advise.

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -287,12 +287,19 @@ void OSDMonitor::update_from_paxos(bool *need_bootstrap)
...
-	dout(10) << " adding osd." << o << " to down_pending_out map" << dendl;
-	down_pending_out[o] = ceph_clock_now(g_ceph_context);
+      if (found == down_pending_out.end()) {
+        dout(10) << " adding osd." << o << " to down_pending_out map" << dendl;
+        down_pending_out[o] = ceph_clock_now(g_ceph_context);
+      }
+    } else {
+      if (found != down_pending_out.end()) {
+        dout(10) << " removing osd." << o << " from down_pending_out map" << dendl;
+        down_pending_out.erase(o);
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: s/o/found/

**Comment by jecluis** [comment]
Would be nice to have a test for this. Other than that, LGTM.

**Comment by tchaikov** [comment]
lgtm.

**Comment by sysnote** [comment]
@jecluis without this PR, i can reproduce everytime though following steps:
1. stop osd.0;
2. start osd.0;
3. wait mon_osd_down_out_interval time, eg:300s;
4. stop osd.0;
5. stop monA(leader);
6. monB won the election, becomes the leader, and then osd.0 is marked out.

with this PR applied, this problem never comes.

ps, I don't have a technology environment, so i just test in my own environment.
Anything i need to do ?

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2016-11-13_05:09:22-rados-wip-kefu-testing---basic-smithi/
- http://pulpito.ceph.com/kchai-2016-11-11_07:45:56-rados-wip-kefu-testing---basic-smithi/

**Comment by tchaikov** [comment]
let's get it into kraken. and add test the test later on.

**Pull request merged**